editor/src/components/editor/persistence/persistence-machine.ts

```diff
@@ -0,0 +1,1216 @@
+            BACKEND_SERVER_SAVE: BackendServerSaving,
+            BACKEND_LOCAL_SAVE: BackendLocalSaving,
+            BACKEND_LOAD: BackendLoading,
+          },
+        },
+        [BackendCreatingProjectId]: {
+          invoke: {
+            id: 'create-project-id',
+            src: getNewProjectId,
+            onDone: {
+              target: BackendIdle,
+              actions: [
+                send((_, event: DoneInvokeEvent<string>) => projectIdCreatedEvent(event.data)),
+              ],
+            },
+            onError: {
+              target: BackendIdle,
+              actions: send((_, event) => backendErrorEvent(event.data)),
+            },
+          },
+        },
+        [BackendCheckingOwnership]: {
+          invoke: {
+            id: 'check-ownership',
+            src: (_, event) => {
+              if (event.type === 'BACKEND_CHECK_OWNERSHIP') {
+                return checkProjectOwned(event.projectId)
+              } else {
+                throw new Error(
+                  `Incorrect event type triggered check ownership, ${JSON.stringify(event)}`,
+                )
+              }
+            },
+            onDone: {
+              target: BackendIdle,
+              actions: [
+                send((_, event: DoneInvokeEvent<boolean>) =>
+                  checkOwnershipCompleteEvent(event.data),
+                ),
+              ],
+            },
+            onError: {
+              target: BackendIdle,
+              actions: send((_, event) => backendErrorEvent(event.data)),
+            },
+          },
+        },
+        [BackendDownloadingAssets]: {
+          invoke: {
+            id: 'download-assets',
+            src: (_, event) => {
+              if (event.type === 'BACKEND_DOWNLOAD_ASSETS') {
+                return downloadAssets(event.projectId, event.projectModel)
+              } else {
+                throw new Error(
+                  `Incorrect event type triggered asset download, ${JSON.stringify(event)}`,
+                )
+              }
+            },
+            onDone: {
+              target: BackendIdle,
+              actions: [
+                send((_, event: DoneInvokeEvent<ProjectWithFileChanges>) =>
+                  downloadAssetsCompleteEvent(event.data),
+                ),
+              ],
+            },
+            onError: {
+              target: BackendIdle,
+              actions: send((_, event) => backendErrorEvent(event.data)),
+            },
+          },
+        },
+        [BackendServerSaving]: {
+          invoke: {
+            id: 'server-save-project',
+            src: (context, event) => {
+              if (
+                event.type === 'BACKEND_SERVER_SAVE' &&
+                event.projectModel != null &&
+                context.projectId != null
+              ) {
+                return saveProjectToServer(context.projectId, event.projectModel)
+              } else {
+                throw new Error(
```

Comment:
What happens to exceptions in the state machine?

Reply:
This will trigger the `onError` below, which then sends an event for the core machine to handle, which right now just squashes them and returns to the "ready" state, so they still need handling before this is complete.